lib/crane/browser/window/scripts/live_view/morphdom.ex

```diff
@@ -0,0 +1,239 @@
+    on_node_discarded = Keyword.get(opts, :on_node_discarded, fn node -> node end)
+    on_before_element_updated = Keyword.get(opts, :on_before_element_updated, fn from_element, _to_element -> from_element end)
+    on_element_updated = Keyword.get(opts, :on_element_updated, fn element -> element end)
+    children_only = Keyword.get(opts, :children_only)
+
+    from_node = GenDOM.Node.get(from_node)
+    to_node = GenDOM.Node.get(to_node)
+
+    zipped_children = :lists.zip(
+      Enum.map(from_node.child_nodes, &GenDOM.Node.get/1),
+      Enum.map(to_node.child_nodes, &GenDOM.Node.get/1),
+      {:pad, {nil, nil}}
+    )
+    Enum.reduce(zipped_children, keyed_nodes, fn
+      {from_child, nil}, acc -> # element removed
+        on_before_node_discarded.(from_child)
+        GenDOM.Node.remove_child(from_node, from_child)
+        on_node_discarded.(from_child)
+        if key = get_node_key.(from_child) do
+          # keep track of keyed nodes, we might match them later
+          Map.put(acc, key, from_child)
+        else
+          acc
+        end
+      {nil, to_child}, acc -> # element added
+        on_before_node_added.(to_child)
+        add_child.(from_node, to_child)
+        on_node_added.(to_child)
+        acc
+      {%{ node_type: @element_node } = from_child, %{ node_type: @element_node } = to_child}, acc -> # element changed
+        if existing_child = Map.get(acc, get_node_key.(to_child)) do
+
+        end
+        on_before_element_updated.(from_child, to_child)
+        Enum.each([:tag_name, :attributes, :class_list], fn prop ->
+          new_value = Map.get(to_child, prop)
+          if Map.get(from_child, prop) != new_value do
+            GenDOM.Node.put(from_child, prop, new_value)
+          end
+        end)
+        on_element_updated.(from_child)
+        morph(
+          from_child,
+          to_child,
+          opts,
+          if key = get_node_key.(from_child) do
+            Map.delete(acc, key)
+          else
+            acc
+          end
+        )
+      {
+        %{ node_type: @text_node, whole_text: from_whole_text } = from_child,
+        %{ node_type: @text_node, whole_text: to_whole_text } = _to_child
+      }, acc when from_whole_text != to_whole_text -> # text changed
+        GenDOM.Node.put(from_child, :whole_text, to_whole_text)
+        acc
+      {%{ node_type: from_node_type } = from_child, %{ node_type: to_node_type } = to_child}, acc when from_node_type != to_node_type -> # text -> element or element -> text
+        on_before_node_discarded.(from_child)
+        on_before_node_added.(to_child)
+        # replace the child
+        GenDOM.Node.replace_child(from_node, to_child, from_child)
+        on_node_discarded.(from_child)
+        on_node_added.(to_child)
+        if key = get_node_key.(from_child) do
+          Map.put(acc, key, from_child)
+        else
+          acc
+        end
+    end)
+  end
+
+  # def morphdom(from_node, to_node, opts \\ []) do
```

Comment:
This commented out section is an implementation I had started on that was more of a clone of the JS, but I abandoned it to get something working in a more functional style.